src/state.rs

```diff
@@ -341,9 +372,39 @@ impl DemandProxyState {
                     .on_demand_dns_cache_misses
                     .get_or_create(&labels)
                     .inc();
-                // TODO: optimize so that if multiple requests to the same hostname come in at the same time,
+                // optimize so that if multiple requests to the same hostname come in at the same time,
                 // we don't start more than one background on-demand DNS task
-                Self::resolve_on_demand_dns(self.to_owned(), workload).await;
+                if state
+                    .get_cached_resolve_dns_for_hostname(&hostname)
+                    .is_none()
+                {
+                    let cached_resolve_dns = SingleFlight::new();
+                    state.set_cached_resolve_dns_for_hostname(
+                        hostname.to_owned(),
+                        cached_resolve_dns,
+                    );
+
+                    let element = state.get_cached_resolve_dns_for_hostname(&hostname);
+                    match element {
+                        Some(sf_map_element) => {
+                            // Check if this is the first request of doing DNS task for hostname
+                            if sf_map_element.calling_index == 1 {
+                                // doing the dns tasks
+                                Self::resolve_on_demand_dns(self.to_owned(), workload).await;
+                                // need to remove the element in sf_by_hostname because the DNS resolving task
+                                // is completed and the IP addresses already are stored in Hashmap of by_hostname
+                                state.remove_cached_resolve_dns_for_hostname(&hostname);
+                                sf_map_element.notify_waiters();
+                            } else {
+                                sf_map_element.wait_for_notifying();
```

Comment:
I am a little concerned a concurrent call for waiting may be after another call notify, in this case one thread may stuck forever

Reply:
Yeah,  I checked it `tokio::sync::Notify` just support the 1 notify vs 1 waiter, I will change it into `tokio::sync::mpsc`.